src/ErrorMessage.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator;
+
+final class ErrorMessage
+{
+    private string $message;
+    /**
+     * @psalm-var array<string, string|integer>
+     */
+    private array $parameters;
+
+    public function __construct(string $message, array $parameters)
+    {
+        $this->message = $message;
+        $this->parameters = $parameters;
+    }
+
+    public function getMessage(): string
+    {
+        return $this->message;
+    }
+
+    public function getParameters(): array
+    {
+        return $this->parameters;
+    }
+
+    public function __toString(): string
```

Comment:
It doesn't do replacement of parameters into message so not the same as ready to use messages.

Reply:
It won't do. Message formatting is a part of `translator` especially `MessageFormatterInterface`.

So if you want either formatted or translated messages you need to add translator as a dependency, configure it and add translator decorator to validator

Reply:
The reason is ability to expose raw error messages with their's parameters. If it can't be, frontend or third-party libraries will get already formatted (with simple formatter or something else) message. For example, `Value 55 is bad` instead of `Value {value} is bad`.
Otherwise it can be solved with saving the original error message that is worse than the current design.